wsbroadcastserver/clientconnection.go

```diff
@@ -76,33 +86,98 @@ func (cc *ClientConnection) Compression() bool {
+		if err != nil {
+			return nil, err
+		}
+
+		data := []byte{}
+		if cc.compression {
+			data = compressed.Bytes()
+		} else {
+			data = notCompressed.Bytes()
+		}
+		err := cc.writeRaw(data)
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("segment sent to client", "client", cc.Name, "sentCount", len(bm.Messages))
+
+		prevSegment = segment
+		segment = segment.Next()
+	}
+	return prevSegment, nil
+
+}
+
 func (cc *ClientConnection) Start(parentCtx context.Context) {
 	cc.StopWaiter.Start(parentCtx, cc)
 	cc.LaunchThread(func(ctx context.Context) {
+		// A delay may be configured, ensures the Broadcaster delays before any
+		// messages are sent to the client. The ClientConnection has not been
+		// registered so the out channel filling is not a concern.
 		if cc.delay != 0 {
-			var delayQueue [][]byte
 			t := time.NewTimer(cc.delay)
-			done := false
-			for !done {
-				select {
-				case <-ctx.Done():
-					return
-				case data := <-cc.out:
-					delayQueue = append(delayQueue, data)
-				case <-t.C:
-					for _, data := range delayQueue {
-						err := cc.writeRaw(data)
-						if err != nil {
-							logWarn(err, "error writing data to client")
-							cc.clientManager.Remove(cc)
-							return
-						}
-					}
-					done = true
-				}
+			select {
+			case <-ctx.Done():
+				return
+			case <-t.C:
 			}
 		}
 
+		// Send the current backlog before registering the ClientConnection in
+		// case the backlog is very large
+		head := cc.backlog.Head()
+		segment, err := cc.writeBacklog(ctx, head)
+		if errors.Is(err, errContextDone) {
+			return
+		} else if err != nil {
+			logWarn(err, "error writing messages from backlog")
+			cc.clientManager.Remove(cc)
+			return
+		}
+
+		cc.clientManager.Register(cc)
+		timer := time.NewTimer(5 * time.Second)
+		select {
+		case <-ctx.Done():
+			return
+		case <-cc.registered:
+			log.Debug("ClientConnection registered with ClientManager", "client", cc.Name)
+		case <-timer.C:
+			log.Warn("timed out waiting for ClientConnection to register with ClientManager", "client", cc.Name)
+		}
+
+		// The backlog may have had more messages added to it whilst the
+		// ClientConnection registers with the ClientManager, therefore the
+		// last segment must be sent again. This may result in duplicate
+		// messages being sent to the client but the client should handle any
+		// duplicate messages. The ClientConnection can not be registered
+		// before the backlog is sent as the backlog may be very large. This
+		// could result in the out channel running out of space.
+		_, err := cc.writeBacklog(ctx, head)
+		if errors.Is(err, errContextDone) {
+			return
+		} else if err != nil {
+			logWarn(err, "error writing messages from backlog")
+			cc.clientManager.Remove(cc)
+			return
+		}
+
+		// TODO(clamb): does this still need to consider the requested seq number from the client? currently it just sends everything in the backlog
```

Comment:
We should use `lookupByIndex` to jump to the requested segment instead of calling `.Head()` (it's ok to replay some messages in the segment before the requested sequence number)